Feature/add permissions tests for controllers

underwater opened the pull request:
Updated the controllers to make use of the HasPermission guard and decorator instead of explicit verification of permissions.  each time.   There were some edge cases, which I've left alone in this initial p/r

- the case DEMO role in updateUserSettings user.controller.ts
- few cases where business rules are also being verified (as per our discussion)
- few cases where multiple permissions are being checked (example: ImportController)  

i suggest to deal with these in a seperate P/R.

dtslvr commented on apps/api/src/app/account-balance/account-balance.controller.ts:
```diff
@@ -10,8 +10,10 @@ import {
 import { REQUEST } from '@nestjs/core';
 import { AuthGuard } from '@nestjs/passport';
-import { AccountBalance } from '@prisma/client';
+import { permissions } from '@ghostfolio/common/permissions';
```
This import seems to be a duplicate

underwater replied:
agree

dtslvr commented on apps/api/src/app/cache/cache.controller.ts:
```diff
@@ -21,19 +15,8 @@ export class CacheController {
 
   @Post('flush')
   @UseGuards(AuthGuard('jwt'))
+  @HasPermission(permissions.accessAdminControl) // permission needed
```
Why this comment?

underwater replied:
forgot to remote it :-)

dtslvr commented on apps/api/src/app/import/import.controller.ts:
```diff
@@ -35,17 +36,14 @@ export class ImportController {
 
   @Post()
   @UseGuards(AuthGuard('jwt'))
+  @HasPermission(permissions.createAccount)
```
I would change this to `permissions.createOrder`

underwater replied:
OK, but then will do the excplicit check on the createAccount (as mentioned previously, I didn't test yet the permission guard / decorator on cases that check multiple permissions. (leaving that for seperate P/R).  OK with you ?

dtslvr commented on apps/api/src/app/import/import.controller.ts:
```diff
@@ -35,17 +36,14 @@ export class ImportController {
-        this.request.user.permissions,
-        permissions.createAccount
-      ) ||
       !hasPermission(this.request.user.permissions, permissions.createOrder)
```
I would change this to `permissions.createAccount`

underwater replied:
OK, got it :-)

dtslvr reviewed (changes requested):
Great how much boilerplate you can eliminate with that 👍🏻

I have added a few commens.

underwater commented:
Fixed to comply with suggestsin....
Also revisited few controllers, and removed unused constructor injection of RequestWithUser and its dependencies where this was no longer needed. So In controllers like account-balance, queue, benchmark, the "@Inject(REQUEST) private readonly request: RequestWithUser" is no longer needed.

dtslvr reviewed (changes requested):
Thank you for the changes, @underwater.
I've tried it, however I always get `403 Forbidden` when I open pages of the _Admin Control_ panel. Other endpoints (non-admin) seems to work, though. Any idea?

underwater commented:
Let me check this tomorrow and revert

On Sun, Dec 24, 2023, 4:05 PM Thomas Kaul ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Thank you for the changes, @underwater <https://github.com/underwater>.
> I've tried it, however I always get 403 Forbidden when I open pages of
> the *Admin Control* panel. Other endpoints (non-admin) seems to work,
> though. Any idea?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ghostfolio/ghostfolio/pull/2771#pullrequestreview-1795593560>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN4N6ZKSMYSFE5LWNCJXVDYLBAELAVCNFSM6AAAAABA6EA3G6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOOJVGU4TGNJWGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dtslvr commented:
> Let me check this tomorrow and revert

Thanks. Do not revert if possible 🙂

underwater commented:
:-) I meant, check and get back to you :-)

On Sun, 24 Dec 2023 at 17:21, Thomas Kaul ***@***.***> wrote:

> Let me check this tomorrow and revert
>
> Thanks. Do not revert if possible 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ghostfolio/ghostfolio/pull/2771#issuecomment-1868552006>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN4N62LUWVKSXW3QMJZJETYLBJBLAVCNFSM6AAAAABA6EA3G6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNRYGU2TEMBQGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

underwater commented:
I suspect to have found the issue

I was registering HasPermissionGuard globally which means it runs earlier in the pipeline than the AuthGuard which is being applied explicitly to each controoler endpoint ... 

As a result HasPermissionGuard was receiving the request before the AuthGuard has had a chance to populate the 'user' object and the HasPermissionGuard was seeing 'user' object as 'undefined'.

To fix this, I would suggest I skip the global registration of the HasPermissionGuard (which I was doing in the app.module.ts) and instead just include it in our @UseGuards(...) like this @UseGuards(AuthGuard('jwt'), HasPermissionGuard)


So to summarize,
1) controller endpoints that do check permissions and indicate the required permissions using @HasPermissionDecorator will need to use @UseGuards(AuthGuard('jwt'), HasPermissionGuard)

while 

2) controller endpoints that don't check permissions (like getPlatforms()) will continue to use the simple
@UseGuards(AuthGuard('jwt'))


To keep things consistent and to avoid confusion, we could also always use the 
@UseGuards(AuthGuard('jwt'), HasPermissionGuard) 

This would work for all cases, the HasPermissionsGuard defaults to allowing the request to pass when no required permission is specified using the @HasPermission(....) 

let me know what is preferred and I will do the chance and push for you to have a look.

underwater commented:
We could eventually also consider registering both guards (AuthGuard, and HasPermissionGuard) globally and in the right order, to avoid having to think about this at each endpoint

dtslvr commented:
Using
```
@UseGuards(AuthGuard('jwt'), HasPermissionGuard)
```
is also fine. To ensure that it is forgotten less often, it may make sense to always use both of them.

underwater commented:
sure, I can update my p/r to make use of @UseGuards(AuthGuard('jwt'), HasPermissionGuard) everywhere for consistency...

underwater commented:
> Using
> 
> ```
> @UseGuards(AuthGuard('jwt'), HasPermissionGuard)
> ```
> 
> is also fine. To ensure that it is forgotten less often, it may make sense to always use both of them.

done ! pushed update

dtslvr reviewed (approved):
Looks fine now. Thanks @underwater 🙂
